module/lifecycle/lifecycle.go

```diff
@@ -49,17 +49,22 @@ func (lm *LifecycleManager) OnStart(startupFns ...func()) {
 // starting up, we will wait for startup to complete before shutting down. After the first
 // call, subsequent calls to OnStop do nothing.
 func (lm *LifecycleManager) OnStop(shutdownFns ...func()) {
+	var waitForStartup bool
+
 	lm.stateTransition.Lock()
 	if lm.shutdownCommenced {
 		lm.stateTransition.Unlock()
 		return
 	}
 	lm.shutdownCommenced = true
+	if lm.startupCommenced {
+		waitForStartup = true
+	}
```

Comment:
Good catch!

wouldn't it be more compact to write:
```suggestion
	waitForStartup := lm.startupCommenced
```
?